### PR TITLE
Improve Buildkite pipeline by addressing feedback from #3758

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR addresses the feedback from #3758.
* a0e6f8fd6020b46c54e0e41e0f17e2b92122259c addresses [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3758#discussion_r677461249)
* 0b97c0226caa0ef66918ee28523427069bfc456c addresses [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3758#discussion_r677470806)
* a85a0487a308ef89f9beb1af8c541b132d1582c3 addresses [this comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3758#discussion_r677168249)

To test:
* Since Buildkite was able to successfully publish the artifacts, we just need to ensure the editor loads without a crash using the [WPAndroid installable build](https://github.com/wordpress-mobile/WordPress-Android/pull/15089#issuecomment-888029063).

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt) if necessary.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -1,13 +1,12 @@
 publish-android-artifacts-docker-container: &publish-android-artifacts-docker-container
-  plugins:
-    - docker#v3.8.0:
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml]: See the comment on the other PR for explanations of how to get the syntax you really wanted here.

TL;DR:
```suggestion
x-common-params:
  &publish-android-artifacts-docker-container
  docker#v3.8.0:
```
Then at call site use:
```yml
plugins:
  - *publish-android-artifacts-docker-container
```

Comment (oguzkocer) [reply .buildkite/pipeline.yml]: Thanks @AliSoftware! I've addressed this in 5e565b7c6f68bfce22036f5cfe2d4f42125e557c and it's looking great now 🥳

File: .buildkite/publish-react-native-aztec-android-artifacts.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -euo pipefail
```
Comment (AliSoftware) [inline .buildkite/publish-react-native-aztec-android-artifacts.sh:3]: Completely opinionated, but as I mentioned to Gio in the other PR, personally I've seen the practice of setting script-wide bash flags like this directly on the shebang instead of using a separate `set` command. Ultimately both do the same, it's just that I've seen the latter used more often in bash script examples and tutorials so considered it might be more common practice? 🤷 

```suggestion
#!/bin/bash -euo pipefail
```

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
-        image: "public.ecr.aws/automattic/android-build-image:v1.0.0"
-        propagate-environment: true
-        environment:
-          # DO NOT MANUALLY SET THESE VALUES!
-          # They are passed from the Buildkite agent to the Docker container
-          - "AWS_ACCESS_KEY"
-          - "AWS_SECRET_KEY"
+x-common-params:
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml:1]: nit and side-note: quick thought that we might want to bikeshed this arbitrary name to come up with a name we all agree with and ultimately try to use the same dummy name for all the YAMLs in all the repositories. `x-common-params`, `common-params`, `yaml-anchors`, `custom-yaml-anchor-definitions`, `reusable-nodes`… there are plenty of ideas and all are valid, so we should just agree on one and stick to it. Seems like Gio ended up using `common-params` in his other PRs, so maybe that'll become our de-facto naming for those?

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
+  &publish-android-artifacts-docker-container
+  docker#v3.8.0:
+    image: "public.ecr.aws/automattic/android-build-image:v1.0.0"
+    propagate-environment: true
+    environment:
+      # DO NOT MANUALLY SET THESE VALUES!
+      # They are passed from the Buildkite agent to the Docker container
+      - "AWS_ACCESS_KEY"
+      - "AWS_SECRET_KEY"
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml:10]: Maybe it's premature "optimization" from my part, but in case we need to add more anchors in the future, I feel like it could make more sense to make the `common-params` dummy dict contain an array (one item per anchor we want to define), rather than directly the sole anchor we currently have. This way if we need to define more anchors as we make the YAML evolve, we can just add another item to the array.

```suggestion
  - &publish-android-artifacts-docker-container
    docker#v3.8.0:
      image: "public.ecr.aws/automattic/android-build-image:v1.0.0"
      propagate-environment: true
      environment:
        # DO NOT MANUALLY SET THESE VALUES!
        # They are passed from the Buildkite agent to the Docker container
        - "AWS_ACCESS_KEY"
        - "AWS_SECRET_KEY"
```

Another option is to make it a dictionary, with the keys being the same keys that we'll use at call site (e.g. `plugins:` for this one), as a kind of reminder of which key this anchor is supposed to be used to in the rest of the YAML.

```suggestion
  plugins: &publish-android-artifacts-docker-container
    docker#v3.8.0:
      image: "public.ecr.aws/automattic/android-build-image:v1.0.0"
      propagate-environment: true
      environment:
        # DO NOT MANUALLY SET THESE VALUES!
        # They are passed from the Buildkite agent to the Docker container
        - "AWS_ACCESS_KEY"
        - "AWS_SECRET_KEY"
```

Though while it helps remind us where this anchor will then be used, that might also bring back the risk of confusion for people not used to anchors in thinking the key name before the anchor would be meaningful while it's not. Maybe a middle ground would be to pick an arbitrary key name prefixed with `x-` (aka: `x-plugins`) or suffixed with `key` (aka: `plugins-key:`), or use `common-plugins` or something like this?

All this is really just a brain dump and total nitpick, just opening the can of worms for the higher discussion, but is not a blocker at all for this PR.

**Review by AliSoftware — APPROVED** [review]
Left a couple of nitpicks and open debate, but totally not blocking.

I think it could at least be worth making `common-params:` be an array to be ready to define more anchors in the future, for the rest (naming of dummy keys etc), it's nitpick and up for wider debate as a light convention we could try to define later.

File: .buildkite/publish-react-native-aztec-android-artifacts.sh (hunk 2 above)
Comment (oguzkocer) [reply .buildkite/publish-react-native-aztec-android-artifacts.sh:3]: Buildkite examples use the extra `set` approach, so I'd prefer to stick with it at least until there is consensus.

File: .buildkite/pipeline.yml (hunk 3 above)
Comment (oguzkocer) [reply .buildkite/pipeline.yml:1]: I've gotten this from your [suggestion here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3764#discussion_r678088237). 🤷 If we want to stick to something specific, let's have that discussion in a p2 with higher visibility. I'd prefer not to keep working on this PR until then.

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (oguzkocer) [reply .buildkite/pipeline.yml:10]: @AliSoftware I think you might be overthinking this as I am having a very hard time following your thought process. The current iteration is exactly what you previously suggested, so I think I'll just keep it as is because it does exactly what I want for the current moment.

I am not saying let's never change it, but when it's laid out like this - especially with the back & forth we already had - it's becoming less constructive. My suggestion is, if you prefer a certain way and you believe there is a good reason for it:

**a.** Open a new PR so the description explains why we made that choice
**b.** Publish an RFC in the internal p2 so we can make a decision and internally document it in the `Automattic/buildkite-ci` so we all follow the same practice.
**c.** Open a PR in `Automattic/buildkite-ci` with YAML best practices, so we can discuss it there and then start applying it everywhere.

Regardless of the option(s) you go with, I'd suggest limiting the alternatives because if we have a lot of them, not only is it confusing and hard to track the thought process, but also it means that there is no clear winner between them.

I hope that makes sense to you!
